chia/server/node_discovery.py

```diff
@@ -291,9 +292,13 @@ async def _connect_to_peers(self, random: Random) -> None:
                         await asyncio.sleep(introducer_backoff)
                     except asyncio.CancelledError:
                         return None
-                    # Run dual between DNS servers and introducers. One time query DNS server,
-                    # next two times query the introducer.
-                    if introducer_attempts % 3 == 0 and len(self.dns_servers) > 0:
+                    # Alternate between DNS servers and introducers.
+                    # First (self.dns_query_ratio - 1) times query the introducer.
+                    # Query the next Chia DNS server in the list once every self.dns_query_ratio attempts.
+                    if (
+                        len(self.dns_servers) > 0
+                        and (introducer_attempts % self.dns_query_ratio) == self.dns_query_ratio - 1
```

Comment:
is the movement of `len(self.dns_servers) > 0` to the front of the expression relevant? is it to save a modulus when we don't have any dns servers?

Other than trying the DNS last, you also change the ratio from 3 to 10. Do these changes all have the objective to reduce load on the introducers?

Reply:
> I don't think this is getting at the functionality change we need. In my tests this appears to just make us fall back and use the dumb old introducer which, frankly, we should rip out (I suspect the only reason we don't is ease of testnet but I digress.)
> 
> With chia.hoffmang.com and dns-introducer.chia.net as my two dns introducers in config.yaml, my node only seems to connect to the old and busted introducer. Optimally we would want to start from the peer db (I have nuked it in my test configuration) and then if we can't get connections go to the list of dns introducers and if we hit ultimate futility - try the old and busted style.

> One of the drivers of this is that the dns introducers do a bit of peer quality checking so the ones they're handing out are known to be pretty good and to have ports available for new incoming connections. Old and busted introducer just dumbly hands out any IP it could connect back to that connected to it at some point with a time decay.

Thanks. I think the issue here is that I'm not understanding the full issue (eg. There is an old, broken introducer, and newer, better introducers -- what is the name of the old introducer?), and also misunderstanding what we are calling "introducers" and "DNS Servers" in the code.

I'll look up the docs on what the intent / desired situation is.

With the incomplete knowledge I have, I'm all for removing the "bad" introducer client side - we can wait to remove the server itself until usage dwindles to nothing.

Reply:
> is the movement of `len(self.dns_servers) > 0` to the front of the expression relevant? is it to save a modulus when we don't have any dns servers?

No, the change has to do with automatic code formatting. I believe this code has the same meaning, even in the face of short-circuiting `and`.

> Other than trying the DNS last, you also change the ratio from 3 to 10. Do these changes all have the objective to reduce load on the introducers?

Yes, that is what I understood from the ticket, but see my above comment to Gene. I was not aware of one of the issues when I put this together.